File: sdk/py/splendor_actor/types/actions.py
from enum import Enum
from typing import Set, Dict, Optional, Union, List

from .color import Color, ColorVec
from .card import Tier

PlayerAction = Union[
    "TakeTokenAction",
    "ReserveCardAction",
    "BuyCardAction",
    "NopAction",
]


class NopAction:
    """
    represent the action of doing nothing
    """

    def to_json(self) -> Dict:
        return {"type": "nop"}


class TakeTokenAction:
    """
    represent the action of taking tokens
    """

    class Type(Enum):
        THREE_DIFFERENT = "three_different"
        TWO_SAME = "two_same"

    _type: Type
    tokens: ColorVec

    def __init__(self, _type: Type, tokens: ColorVec):
        self._type = _type
        self.tokens = tokens

    @classmethod
    def three_different(cls, tokens: Union[Set[Color], List[Color]]):
        """
        create a TakeTokenAction with at most three different tokens
        """
        assert len(tokens) <= 3
        vec = ColorVec.empty()
        if isinstance(tokens, list):
            tokens = set(tokens)
        for token in tokens:
            vec[token] = 1
        return cls(cls.Type.THREE_DIFFERENT, vec)

    @classmethod
    def two_same(cls, token: Color):
        """
        create a TakeTokenAction with two same tokens
        """
        tokens = ColorVec.empty()
        tokens[token] = 2
        return cls(cls.Type.TWO_SAME, tokens)

    def to_json(self) -> Dict:
        return {
            "type": "take_tokens",
            "action": {"type": self._type.value, "tokens": self.tokens.to_json()},
        }


class ReserveCardAction:
    """
    represent the action of reserving a card
    """

    def __init__(self, tier: Tier, idx: Optional[int]):
        self.tier = tier
        self.idx = idx

    @classmethod
    def from_revealed(cls, tier: Tier, idx: int):
        """
        create a ReserveCardAction from the revealed cards
        """
        assert 0 <= idx < 4

        return cls(tier, idx)

    @classmethod
    def from_pool(cls, tier: Tier):
        """
        create a ReserveCardAction from the card pool
        """
        return cls(tier, None)

    def to_json(self) -> Dict:
        return {
            "type": "reserve_card",
            "action": {
                "tier": self.tier.value,
                "idx": self.idx,
            },
        }


class BuyCardAction:
    """
    represent the action of buying a card
    """

    class SourceType(Enum):
        REVEALED = "revealed"
        RESERVED = "reserved"

    class RevealedCardLocation:
        tier: Tier
        idx: int

        def __init__(self, tier: Tier, idx: int):
            self.tier = tier
            self.idx = idx

        def to_json(self) -> Dict:
            return {
                "tier": self.tier.value,
                "idx": self.idx,
            }

    class ReservedCardLocation:
        idx: int

        def __init__(self, idx: int):
            self.idx = idx

        def to_json(self) -> int:
            return self.idx

    source_type: SourceType
    location: Union[RevealedCardLocation, ReservedCardLocation]
    uses: ColorVec

    def __init__(
        self,
        source_type: SourceType,
        location: Union[RevealedCardLocation, ReservedCardLocation],
        uses: ColorVec,
    ):
        self.source_type = source_type
        self.location = location
        self.uses = uses

    @classmethod
    def from_revealed(cls, tier: Tier, idx: int, uses: ColorVec):
        """
        create a BuyCardAction from the revealed cards
        """
        assert 0 <= idx < 4

        return cls(cls.SourceType.REVEALED, cls.RevealedCardLocation(tier, idx), uses)

    @classmethod
    def from_reserved(cls, idx: int, uses: ColorVec):
        """
        create a BuyCardAction from the reserved cards
        """
        assert 0 <= idx < 3
        return cls(cls.SourceType.RESERVED, cls.ReservedCardLocation(idx), uses)

    def to_json(self) -> Dict:
        return {
            "type": "buy_card",
            "action": {
                "source": {
                    "type": self.source_type.value,
                    "location": self.location.to_json(),
                },
                "uses": self.uses.to_json(),
            },
        }


class DropTokensAction:
    """
    represent the action of dropping tokens
    """

    tokens: ColorVec

    def __init__(self, tokens: ColorVec):
        self.tokens = tokens

    def to_json(self) -> List[int]:
        return self.tokens.to_json()


class SelectNoblesAction:
    """
    represent the action of selecting nobles
    """

    idx: int

    def __init__(self, idx: int):
        assert 0 <= idx < 5
        self.idx = idx

    def to_json(self) -> int:
        return self.idx

File: sdk/go/pkg/types/actions.go
package types

import "encoding/json"

type playerActionType string
type takeTokenActionType string
type buyCardSourceType string

const (
	nopAction         playerActionType = "nop"
	takeTokenAction   playerActionType = "take_tokens"
	reserveCardAction playerActionType = "reserve_card"
	buyCardAction     playerActionType = "buy_card"

	takeThreeDifferentToken takeTokenActionType = "three_different"
	takeTwoSameToken        takeTokenActionType = "two_same"

	revealedCard buyCardSourceType = "revealed"
	reservedCard buyCardSourceType = "reserved"
)

type PlayerAction interface {
	GetType() playerActionType
	MarshalJSON() ([]byte, error)
}

type NopAction struct{}

func (a NopAction) GetType() playerActionType {
	return nopAction
}

func (a NopAction) MarshalJSON() ([]byte, error) {
	return json.Marshal(struct {
		Type playerActionType `json:"type"`
	}{
		Type: nopAction,
	})
}

type TakeTokenAction struct {
	Type   takeTokenActionType
	Tokens ColorVec
}

func (a TakeTokenAction) GetType() playerActionType {
	return takeTokenAction
}

func (a TakeTokenAction) MarshalJSON() ([]byte, error) {
	action := struct {
		Type   takeTokenActionType `json:"type"`
		Tokens ColorVec            `json:"tokens"`
	}{
		Type:   a.Type,
		Tokens: a.Tokens,
	}
	marshal := struct {
		Type   playerActionType `json:"type"`
		Action interface{}      `json:"action"`
	}{
		Type:   takeTokenAction,
		Action: action,
	}
	return json.Marshal(marshal)
}

// create a TakeTokenAction with at most three different tokens
func NewTakeThreeDifferentTokenAction(tokens []Color) TakeTokenAction {
	vec := ColorVec{}
	for _, c := range tokens {
		vec[c] = 1
	}
	return TakeTokenAction{
		Type:   takeThreeDifferentToken,
		Tokens: vec,
	}
}

// create a TakeTokenAction with two same tokens
func NewTakeTwoSameTokenAction(token Color) TakeTokenAction {
	vec := ColorVec{}
	vec[token] = 2
	return TakeTokenAction{
		Type:   takeTwoSameToken,
		Tokens: vec,
	}
}

type ReserveCardAction struct {
	Tier Tier
	Idx  *uint8
}

func (a ReserveCardAction) GetType() playerActionType {
	return reserveCardAction
}

func (a ReserveCardAction) MarshalJSON() ([]byte, error) {
	action := struct {
		Tier Tier   `json:"tier"`
		Idx  *uint8 `json:"idx"`
	}{
		Tier: a.Tier,
		Idx:  a.Idx,
	}
	marshal := struct {
		Type   playerActionType `json:"type"`
		Action interface{}      `json:"action"`
	}{
		Type:   reserveCardAction,
		Action: action,
	}
	return json.Marshal(marshal)
}

// NewReserveCardFromRevealedAction creates a ReserveCardAction from the revealed cards
func NewReserveCardFromRevealedAction(tier Tier, idx uint8) ReserveCardAction {
	return ReserveCardAction{
		Tier: tier,
		Idx:  &idx,
	}
}

// NewReserveCardFromPoolAction creates a ReserveCardAction from the card pool
func NewReserveCardFromPoolAction(tier Tier) ReserveCardAction {
	return ReserveCardAction{
		Tier: tier,
		Idx:  nil,
	}
}

type buyCardActionSource struct {
	Type     buyCardSourceType `json:"type"`
	Location interface{}       `json:"location"`
}

type BuyCardAction struct {
	Source buyCardActionSource `json:"source"`
	Uses   ColorVec            `json:"uses"`
}

func (a BuyCardAction) GetType() playerActionType {
	return buyCardAction
}

func (a BuyCardAction) MarshalJSON() ([]byte, error) {
	action := struct {
		Source buyCardActionSource `json:"source"`
		Uses   ColorVec            `json:"uses"`
	}{
		Source: a.Source,
		Uses:   a.Uses,
	}
	marshal := struct {
		Type   playerActionType `json:"type"`
		Action interface{}      `json:"action"`
	}{
		Type:   buyCardAction,
		Action: action,
	}
	return json.Marshal(marshal)
}

// NewBuyCardFromRevealedAction creates a BuyCardAction from the revealed cards
func NewBuyCardFromRevealedAction(tier Tier, idx uint8, uses ColorVec) BuyCardAction {
	return BuyCardAction{
		Source: buyCardActionSource{
			Type: revealedCard,
			Location: struct {
				Tier Tier  `json:"tier"`
				Idx  uint8 `json:"idx"`
			}{
				Tier: tier,
				Idx:  idx,
			},
		},
		Uses: uses,
	}
}

// NewBuyCardFromReservedAction creates a BuyCardAction from the reserved cards
func NewBuyCardFromReservedAction(idx uint8, uses ColorVec) BuyCardAction {
	return BuyCardAction{
		Source: buyCardActionSource{
			Type:     reservedCard,
			Location: idx,
		},
		Uses: uses,
	}
}
